packages/beacon-node/src/sync/unknownBlock.ts

```diff
@@ -825,4 +753,231 @@ export class UnknownBlockSync {
+   * make this public for testing
+   */
+  onRequest(peerId: PeerIdStr): void {
+    if (this.activeRequests.has(peerId)) {
+      this.activeRequests.set(peerId, (this.activeRequests.get(peerId) ?? 0) + 1);
+    } else {
+      this.activeRequests.set(peerId, 1);
+    }
+  }
+
+  /**
+   * Consumers should call this method when a request is completed for a peer.
+   */
+  onRequestCompleted(peerId: PeerIdStr): void {
+    if (this.activeRequests.has(peerId)) {
+      const count = this.activeRequests.get(peerId) ?? 0;
+      this.activeRequests.set(peerId, Math.max(0, count - 1));
+    }
+  }
+
+  /**
+   * For testing only
+   */
+  getActiveRequest(peerId: PeerIdStr): number {
+    return this.activeRequests.get(peerId) ?? 0;
+  }
+
+  getPeersMetaCount(): number {
+    return this.peersMeta.size;
+  }
+
+  getPeersActiveRequestCount(): number {
+    return this.activeRequests.size;
+  }
+
+  getTotalActiveRequests(): number {
+    let totalActiveRequests = 0;
+    for (const count of this.activeRequests.values()) {
+      totalActiveRequests += count;
+    }
+    return totalActiveRequests;
+  }
+
+  // pendingDataColumns could be null for prefulu
+  private filterPeers(pendingDataColumns: Set<number> | null, excludedPeers: Set<PeerIdStr>): PeerIdStr[] {
+    let maxColumnCount = 0;
+    const considerPeers: {peerId: PeerIdStr; columnCount: number}[] = [];
+    for (const [peerId, syncMeta] of this.peersMeta.entries()) {
+      if (excludedPeers.has(peerId)) {
+        // this peer is requested already
+        continue;
+      }
+
+      const activeRequests = this.activeRequests.get(peerId) ?? 0;
+      if (activeRequests >= MAX_CONCURRENT_REQUESTS) {
+        // should return peer with no more than MAX_CONCURRENT_REQUESTS active requests
+        continue;
+      }
+
+      if (pendingDataColumns === null || pendingDataColumns.size === 0) {
+        // prefulu, no pending columns
+        considerPeers.push({peerId, columnCount: 0});
+        continue;
+      }
+
+      // postfulu, find peers that have custody columns that we need
+      const {custodyGroups: peerColumns} = syncMeta;
+      // check if the peer has all needed columns
+      const neededColumns = this.custodyConfig.sampledColumns.reduce((acc, elem) => {
+        if (pendingDataColumns.has(elem)) {
+          acc.push(elem);
+        }
+        return acc;
+      }, [] as number[]);
+
+      // get match
+      const columns = peerColumns.reduce((acc, elem) => {
+        if (neededColumns.includes(elem)) {
+          acc.push(elem);
+        }
+        return acc;
+      }, [] as number[]);
+
+      if (columns.length > 0) {
+        if (columns.length > maxColumnCount) {
+          maxColumnCount = columns.length;
+        }
+        considerPeers.push({peerId, columnCount: columns.length});
+      }
+    } // end for
+
+    const eligiblePeers: PeerIdStr[] = [];
+    for (const {peerId, columnCount} of considerPeers) {
+      if (columnCount === maxColumnCount) {
```

Comment:
We would only make a peer eligible if they have the maxColumnCount that matches other peers?  Seems like we would want to take others with smaller overlap also incase the ones with max are already busy with other requests.